File: KLASA_3/ROZDZIAL_1/LEKCJA_1/PYTHON/ZAD8.py
from stos import Stos


def czy_para(otwierajacy, zamykajacy):
    pary = {
        '(': ')',
        '{': '}',
        '[': ']'
    }
    return pary[otwierajacy] == zamykajacy


def czy_wiekszy(nawias1, nawias2):
    moce = {
        '(': 0,
        '[': 1,
        '{': 2
    }
    return moce[nawias1] > moce[nawias2]


def sprawdz_wyrazenie(wyrazenie_nawiasowe):
    stos = Stos()
    for nawias in wyrazenie_nawiasowe:
        if nawias in '({[':
            if not stos.isEmpty():
                ostatni = stos.peek()
                if czy_wiekszy(nawias, ostatni):
                    return False
            stos.push(nawias)
        elif nawias in ')}]':
            if stos.isEmpty():
                return False
            else:
                otwierajacy = stos.peek()
                stos.pop()
                if not czy_para(otwierajacy, nawias):
                    return False
    return stos.isEmpty()


print('TAK' if sprawdz_wyrazenie(input('Podaj wyrażenie nawiasowe: ')) else 'NIE')

File: KLASA_3/ROZDZIAL_1/LEKCJA_1/C++/ZAD8.cpp
#include <iostream>
#include <stack>
#include <map>

using namespace std;

bool czy_para(char otwierajacy, char zamykajacy)
{
    map<char,char> pary;
    pary['('] = ')';
    pary['['] = ']';
    pary['{'] = '}';

    return pary[otwierajacy] == zamykajacy;
}

bool czy_wiekszy(char nawias1, char nawias2)
{
    map<char,int> moce;
    moce['('] = 0;
    moce['['] = 1;
    moce['{'] = 2;
    return moce[nawias1] > moce[nawias2];
}

bool sprawdz_wyrazenie(string wyrazenie)
{
    stack<char> stos;

    for(char nawias: wyrazenie){
        if(nawias=='(' || nawias=='[' || nawias=='{'){
            if(!stos.empty()){
                char ostatni = stos.top();
                if(czy_wiekszy(nawias,ostatni))return false;
            }
            stos.push(nawias);
        }else{
            if(stos.empty())return false;
            char otwierajacy = stos.top();
            stos.pop();
            if(!czy_para(otwierajacy, nawias))return false;
        }
    }

    return stos.empty();
}

int main()
{
    string wyrazenie;

    cout << "Podaj wyrazenie nawiasowe: ";
    cin >> wyrazenie;

    cout << (sprawdz_wyrazenie(wyrazenie) ? "TAK" : "NIE");

    return 0;
}
